Branch predictor implementation

jiristefan opened the pull request:
Resolves #64
Should also address the branch predictor TODOs from #1 and #2

This PR implements:
- Branch Target Table (BTB) for storing instruction address / target address pairs for jump and branch instruction, along with a UI dock widgets for visualizing the table
- 3 simple static branch predictors (Always Taken / Not Taken, Backward Taken Forward Not Taken)
- 3 Smith predictors (1-bit, 2-bit, 2-bit with hysteresis), along with a Branch History Table (BHT)
- Variable BTB and BHT sizes, from 0 to 8 bits (0 to 16 bits for BHT, might decrease to 8 in the future)
- UI dock widget for visualizing the BHT contents, predictions, and predictor statistics
- Optional Branch History Register for addressing the BHT

Predictor info widget with BHT:
![bht-prediction](https://github.com/cvut/qtrvsim/assets/100444966/ad200d32-07c2-4818-91fc-259784d1ef5e)

BTB widget:
![btb-prediction](https://github.com/cvut/qtrvsim/assets/100444966/5037bf67-ca2e-41c4-877f-ae0028351819)

Configuration dialog:
![qtrvsim-newdialog](https://github.com/cvut/qtrvsim/assets/100444966/80fd5e83-e061-4703-8c72-b44470cc9fd2)

jdupak reviewed (commented):
@ppisa I have found no major issue with the code. It could be simplified a lot and I don't like the repetition of information in many places, but it can be accepted and improved later.

ppisa commented:
@jiristefan please focus on the thesis now and then follow @jdupak  suggestions. When you push to your branch the pull request should automatically update as well.

I have tested next code on your branch 233b331d5c40183a7abd2ebc17e94af35a348d12 with sequence tuned to have no aliases and results seems to be correct

```
   addi a0, zero, 10
l: nop
   j a
   nop
a: nop
   bne zero, zero, b
   nop
   nop
b: beq zero, zero, c
   nop
c: addi a0, a0, -1
   bne a0, zero, l
   ebreak 
```

with Smith 1 bit BTB bits 3, BHR 0, BHT 2, initial NT  and results are strange on single-cycle and pipelined and results seems to be correct.

When I add BHR 2, I get correct 35, incorrect 5 accuracy 87%.

But when I switch to pipelined I get correct 28, incorrect 12 accuracy 70%.

Adding BHR support above the assignment you have complicated things for yourselves. It is nice to have it there, but you probably need to pass through pipeline the index of the entry used at the prediction time because BHR updates when some branches are near. Even this solution would have problems depending on the variable length stalls caused by possible different cache hit miss in the sequences. Using local history would be easier there probably a little. But it can be added in future to the selection of the Smitch etc...

I do not see there problems with global BHR as blocker, it should be documented that for precise expected behavior used in the textbooks the core has to be used in the single-cycle variant. But some analysis how to resolve problem for pipelined and even for superscalar cores should be considered to update behavior in future.

jiristefan commented:
Thank you for the review. I will fix the highlighted issues during next week, I admit the frontend code is more messy than the internal implementation, I'll try to implement all the suggestions and generally clean it up.

As for the branch history register, I will probably use the suggested approach of storing the index with the instruction address locally in the predictor, I believe it should not take too much time to implement properly.

ppisa commented:
@jiristefan I have solution to make BHR work correctly for pipelined version. You need two copies of BHR. The one used to form index at PC computation stage is updated from outcomes of predictions and then the second one is updated in MEM stage from actual results of the branch condition evaluations. They will be in delayed sync which will be maintained as long as there is no missprediction. At the flush event MEM stage BHR will be copied to PC stage BHR. This way the predictor should work exactly the same way/with same outcome for singlecycle and pipelined version. So when you submit the text you can try to implement this approach.

jiristefan commented on .gitignore:
```diff
@@ -11,8 +11,25 @@
+Makefile
+cmake_install.cmake
+CTestTestfile.cmake
+CMakeCache.txt
```
I'm now (hopefully) properly using the cmake command and building the project in another folder. Reverted the commit.

jiristefan commented on src/gui/dialogs/new/newdialog.cpp:
```diff
@@ -365,6 +385,102 @@ void NewDialog::reset_at_compile_change(bool v) {
+        ui->text_bp_bhr_bits_number->setEnabled(true);
+        ui->slider_bp_bht_addr_bits->setEnabled(true);
+        ui->text_bp_bht_addr_bits_number->setEnabled(true);
+        ui->text_bp_bht_bits_number->setEnabled(true);
```
Created new function `bp_toggle_history_table_ui()` to handle this.

jiristefan commented on src/gui/dialogs/new/newdialog.cpp:
```diff
@@ -365,6 +385,102 @@ void NewDialog::reset_at_compile_change(bool v) {
+    };
+
+    if (config->get_bp_type() == machine::PredictorType::SMITH_1_BIT) {
+        ui->select_bp_init_state->addItem(
```
Moved hopefully all the strings into functions for converting the enum value into a string.
There are three now, one for branch result, one for predictor type and one fot its state.

I considered assigning the string to the enum values, but I'm not sure using those string values for comparison is the right way to go, and the functions allow me to return shortened versions of the strings as well.

Let me know if there is a better way to do this.

jiristefan commented on src/gui/dialogs/new/newdialog.cpp:
```diff
@@ -365,6 +385,102 @@ void NewDialog::reset_at_compile_change(bool v) {
+}
+
+void NewDialog::bp_init_state_change() {
+    config->set_bp_init_state(
```
Personally, I think it would make the code less readable? Some of the functions for handling the GUI would have their own declaration, while other functions which happen to be shorter would be defined right in the connect() function call in the place of the slot. Is that correct?

If that is fine and the intended way this should be written, then I'll rewrite it, I just wanted to check.

jdupak replied:
I mean, this is not some exposed functionality, this is purely internal to the component. And having whole clutter of function declaration and everything just to say, "if this changes, enable this" does not seems useful. I prefer the internal behavior to be in one place, not 3 (declaration, definition, connect).
I will leave it to you, but I do this a lot in other parts of the gui.

jiristefan replied:
I've made most of the slots into lambdas, as you suggested, so this should be addressed now.

jiristefan commented on src/gui/dialogs/new/newdialog.cpp:
```diff
@@ -379,6 +495,38 @@ void NewDialog::config_gui() {
+        ui->select_bp_type->removeItem(0);
+    }
+    ui->select_bp_type->addItem(
+        "Always not taken", QVariant((uint8_t)machine::PredictorType::ALWAYS_NOT_TAKEN));
```
Should also be addressed by moving those strings to a function for converting the enum.

jiristefan commented on src/gui/mainwindow/mainwindow.h:
```diff
@@ -94,6 +96,11 @@ public slots:
     void reset_windows();
     void show_symbol_dialog();
+    // Branch predictor
+    void show_dock_bp_btb();
```
Fixed

jiristefan commented on src/gui/windows/predictor/predictor_bht_dock.h:
```diff
@@ -0,0 +1,124 @@
+    uint8_t number_of_bhr_bits;
+    uint8_t number_of_bht_bits;
+
+    QGroupBox *content;
```
Fixed

jiristefan commented on src/gui/windows/predictor/predictor_btb_dock.cpp:
```diff
@@ -0,0 +1,163 @@
+
+uint8_t DockPredictorBTB::init_number_of_bits(const uint8_t b) const {
+    if (b > BP_MAX_BTB_BITS) {
+        WARN("Number of BTB bits (%u) was larger than %u during init", b, BP_MAX_BTB_BITS);
```
The function was not used anywhere, probably copied it from the branch predictor class earlier in development. I removed it.

jiristefan commented on src/machine/predictor.h:
```diff
@@ -1,23 +1,332 @@
-class Predictor {
-public:
-    virtual Address predict(Instruction inst, Address addr) = 0;
+const char *int_to_bit_string(const uint64_t value, const uint8_t size = 8);
```
I'm not sure how the QStringView should be implemented correctly, to be honest. This implementation allows the returned value of the function to be used in all needed places without further processing.

With QStringView as return value, I need to wrap all strings in QStringViewLiteral() macro (am I using that correctly?) and convert the return value to QString, using the .toString() method, to use the value pointed at by QStringView in any of the Qt UI widgets. Additionally for the LOG() function, I need to convert the QString using qPrintable() macro to make it work withou warning or error.

For now I've implemented it like this, but please let me know the correct usage / where can I read about it.

jdupak replied:
> With QStringView as return value, I need to wrap all strings in QStringViewLiteral() macro

Not really, just make them UTF-16 literals. `u"Hello"` Then it will work fine. Constructing the QStringView itself is cheap.

> Additionally for the LOG() function, I need to convert the QString using qPrintable() macro to make it work withou warning or error.

Yes, but that is nothing new. We are using raw C strings practically nowhere (I hope).

> convert the return value to QString, using the .toString() method

Maybe. Some places accept string view (inside qt), some do not. To be clear, I would also accept returning a QString. But not raw pointers.

jiristefan replied:
Understood. I fixed that, and will leave it with the QStringView as a return value, if that is preferred.

jiristefan commented on src/machine/predictor.h:
```diff
@@ -1,23 +1,332 @@
+
+const char *predictor_state_to_string(const PredictorState state);
+
+QString addr_to_hex_str(const machine::Address address);
```
In rest of the project I found this is done in many places using some variation of this code:
`
QString s, t;
t = QString::number(address.get_raw(), 16);
s.fill('0', 8 - t.count());
return "0x" + s + t;
`
I've just taken this code and put it inside a function for my convenience, as I need to do this in four places in the branch predictor UI. I didn't know wehere to put this utility function, and so I just placed it in the file which I import in the UI with the rest of the predictor code.

I think ideally, the Address class would have a member function `to_hex_str()` or similar, which would return the result of the above mentioned code snipped, and that function would then  be used in all places this is needed. That would seem much simpler to me.

jiristefan commented:
I should mention that I moved the contents of the "Core" tab in the dialog window for configuring the simulation into a scroll area widget:
![image](https://github.com/cvut/qtrvsim/assets/100444966/8f13ee38-5d47-40c2-a6d5-7fade5815eba)

The branch predictor config added a bit of height to the tab, and there are couple more settings I would like to add to the tab regarding the predictor in the future. So, to prevent the window from getting too tall, I moved everything to the scroll area. Hopefully there will be no issue with this.

jiristefan commented:
@jdupak I think the reason the Qt6 check is failing is probably because of the predictor_types.h file which I added with the predictor enums I use in the code. To use the enums inside QVariant in the UI objects, I had to register the enum using the provided macros Q_NAMESPACE and Q_ENUM_NS.

I got a very similar error with Qt5 when I tried to compile the code after that, before noticing I did not add the predictor_types.h to the CMakeLists.txt file, then it worked fine.

I'm not sure what changed in Qt6, or if this is a good way to handle this, so please let me know how to fix this. An alternative would be to just convert the enums to integers before storing them in the QVariant, and then back when reading them.

ppisa commented:
The branch prediction demonstration developed by Jiri Stefan in the frame of his master's thesis is welcomed step for QtRvSim to cover more computer architectures lectures topics. The visualization and correct/wrong prediction statistic counting matches teaching needs of the classes when demonstrated on single-cycle processor. For pipelined version there are more topics to discussion, see https://github.com/cvut/qtrvsim/issues/143. It is questionable if solution matching basic textbooks principles on the pipelined version without classification of branched during cache fills can be found. Probably not, but some updates would provide better insight to the problem and code should be enhanced.
